### PR TITLE
Add `FallibleIterator` extension trait

### PR DESCRIPTION
**Pull request by TylerBloom**
This PR adds a `utils` module at the root of apollo federation. In it is an `Iterator` extension trait to help work with iterators over `Result`s, operations that yield `Result`s, and both at the same time.

The extension trait is by no means complete or designed to cover every usecase. The intention is to cover common usecases now and expand things as they are needed. Of particular interest for me were variant of `any`, `all`, and `filter` (plus `and_then`).

### CODE REVIEW

**Comment by TylerBloom** [comment]
I am still working on docs and tests.

**Comment by TylerBloom** [comment]
While working on this, it occurred to me that there is nothing specific to Apollo about this. Perhaps this could/should be mirrored and published as a separate crate? I don't know what Apollo's policy is on that.

File: apollo-federation/src/utils/fallible_iterator.rs [hunk 1]
```diff
@@ -0,0 +1,344 @@
+#![deny(rustdoc::broken_intra_doc_links)]
+
+use itertools::Itertools;
+
+#[allow(dead_code, unused)]
```
Comment (SimonSapin) [inline apollo-federation/src/utils/fallible_iterator.rs:5]: Why not `#[test]` instead?

**Review by SimonSapin — COMMENTED** [review]
Is there nothing at all in the rest of the code base that can use these utilities? Showing actual use in this PR would help motivate their existence. I’m a bit reluctant to add them "just in case"

**Comment by goto-bus-stop** [comment]
I agree it would be good to update some usage sites in a separate commit.

Is there a reason you picked `fallible_*` over `try_*`? I think a `.try_filter()` that short-circuits when the predicate returns Err is perfectly in line with `.try_collect()` in itertools

**Comment by lrlna** [comment]
@goto-bus-stop @SimonSapin the two of you had comments previously on this PR, do you have time in the next week to review again, or discuss with @TylerBloom if this is something we'd like in the repo? I don't want this branch to get too stale.

File: apollo-federation/src/operation/mod.rs [hunk 2]
```diff
@@ -2325,12 +2327,14 @@ impl SelectionSet {
...
-            if selection.conditions()? != conditions {
-                return Ok(Conditions::Boolean(true));
-            }
+        if selections
+            .map(|selection| selection.conditions())
+            .ok_and_any(|cond| cond != conditions)?
+        {
+            Ok(Conditions::Boolean(true))
+        } else {
+            Ok(conditions)
```
Comment (SimonSapin) [inline apollo-federation/src/operation/mod.rs:2336]: I’m not super convinced this is an improvement over the `for` loop but 🤷

File: apollo-federation/src/query_graph/mod.rs [hunk 3]
```diff
@@ -901,17 +899,15 @@ impl QueryGraph {
...
-
-        Ok(false)
+        schema
+            .possible_runtime_types(interface_field_definition_position.parent().into())?
+            .into_iter()
+            .map(|object_type_definition_position| {
+                object_type_definition_position
+                    .field(interface_field_definition_position.field_name.clone())
+            })
+            .map(|field_pos| field_pos.get(schema.schema()))
```
Comment (SimonSapin) [inline apollo-federation/src/query_graph/mod.rs:909]:
```suggestion
            .map(|object_type_definition_position| {
                let field_pos = object_type_definition_position
                    .field(interface_field_definition_position.field_name.clone());
                field_pos.get(schema.schema())
            })
```

File: apollo-federation/src/query_graph/mod.rs [hunk 4]
```diff
@@ -901,17 +899,15 @@ impl QueryGraph {
...
+            .ok_and_any(|field| field.directives.has(&provides_directive_definition.name))
+            .map_err(Into::into)
```
Comment (SimonSapin) [inline apollo-federation/src/query_graph/mod.rs:911]: Is there a nice way to replace this map_err with a `?` operator?

File: apollo-federation/src/supergraph/mod.rs [hunk 5]
```diff
@@ -96,16 +98,19 @@ pub(crate) fn extract_subgraphs_from_supergraph(
...
+        .fallible_filter(|type_definition_position| {
+            join_spec_definition
+                .is_spec_type_name(supergraph_schema, type_definition_position.type_name())
+                .map(Not::not)
+        })
+        .and_then_filter(|type_definition_position| {
+            link_spec_definition
+                .is_spec_type_name(supergraph_schema, type_definition_position.type_name())
+                .map(Not::not)
+        })
```
Comment (SimonSapin) [inline apollo-federation/src/supergraph/mod.rs:112]: `.map(not)` feels very awkward to me, I think I prefer `?` +  `Ok`:
```suggestion
        .fallible_filter(|type_definition_position| {
            Ok(!join_spec_definition
                .is_spec_type_name(supergraph_schema, type_definition_position.type_name())?)
        })
        .and_then_filter(|type_definition_position| {
            Ok(!link_spec_definition
                .is_spec_type_name(supergraph_schema, type_definition_position.type_name())?)
        })
```

**Review by SimonSapin — APPROVED** [review]
Approved with `pub(crate)`, the rest as you see fit

File: apollo-federation/src/query_graph/mod.rs (hunk 4 above)
Comment (TylerBloom) [reply apollo-federation/src/query_graph/mod.rs:911]: `map_err(Into::into)` can be replaced with `?` and wrapping the whole express in an `Ok`. I generally dislike `Ok(fallible()?)`, but that's kind of bikeshedding. It doesn't really matter.

File: apollo-federation/src/operation/mod.rs (hunk 2 above)
Comment (TylerBloom) [reply apollo-federation/src/operation/mod.rs:2336]: Ya, I was probably a little overly zealous here.

File: apollo-federation/src/supergraph/mod.rs (hunk 5 above)
Comment (TylerBloom) [reply apollo-federation/src/supergraph/mod.rs]: I agree with you, but there is a reason for this. The compiler can not infer what the error type is for the `Ok`s. This is true with the `fallible_filter` + `and_then_filter` and if the two filters combined into something like:
```rs
        .fallible_filter(|type_definition_position| {
            Ok(!join_spec_definition
                     .is_spec_type_name(supergraph_schema, type_definition_position.type_name())?
                  ||  !link_spec_definition
                        .is_spec_type_name(supergraph_schema, type_definition_position.type_name())?
        )})
```

**Pull request merged**